internal/overlord/servstate/handlers.go

```diff
@@ -51,183 +54,528 @@ var (
 		}
 	}
+}
 
-	args, err := shlex.Split(service.Command)
+func (m *ServiceManager) removeService(name string) {
+	m.servicesLock.Lock()
+	delete(m.services, name)
+	m.servicesLock.Unlock()
+}
+
+// transition changes the service's state machine to the given state.
+func (s *service) transition(state serviceState) {
+	logger.Debugf("Service %q transitioning to state %q", s.config.Name, state)
+	s.state = state
+}
+
+func (s *service) getStateLocked() serviceState {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	return s.state
+}
+
+// start is called to transition from the initial state and start the service.
+func (s *service) start() error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	switch s.state {
+	case stateInitial:
+		err := s.startInternal()
+		if err != nil {
+			return err
+		}
+		s.transition(stateStarting)
+		time.AfterFunc(okayWait, func() { logError(s.okayWaitElapsed()) })
+
+	default:
+		return fmt.Errorf("start invalid in state %q", s.state)
+	}
+	return nil
+}
+
+func logError(err error) {
+	if err != nil {
+		logger.Noticef("%s", err)
+	}
+}
+
+// startInternal is an internal helper used to actually start (or restart) the
+// command. It assumes the caller has ensures the service is in a valid state,
+// and it sets s.cmd and other relevant fields.
+func (s *service) startInternal() error {
+	args, err := shlex.Split(s.config.Command)
 	if err != nil {
 		// Shouldn't happen as it should have failed on parsing, but
 		// it does not hurt to double check and report.
 		return fmt.Errorf("cannot parse service command: %s", err)
 	}
-	cmd := exec.Command(args[0], args[1:]...)
-	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
+	s.cmd = exec.Command(args[0], args[1:]...)
+	s.cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
 
-	// Start as another user if specified in plan
-	uid, gid, err := osutil.NormalizeUidGid(service.UserID, service.GroupID, service.User, service.Group)
+	// Start as another user if specified in plan.
+	uid, gid, err := osutil.NormalizeUidGid(s.config.UserID, s.config.GroupID, s.config.User, s.config.Group)
 	if err != nil {
 		return err
 	}
 	if uid != nil && gid != nil {
-		setCmdCredential(cmd, &syscall.Credential{
+		setCmdCredential(s.cmd, &syscall.Credential{
 			Uid: uint32(*uid),
 			Gid: uint32(*gid),
 		})
 	}
 
-	// Pass service description's environment variables to child process
-	cmd.Env = os.Environ()
-	for k, v := range service.Environment {
-		cmd.Env = append(cmd.Env, k+"="+v)
+	// Pass service description's environment variables to child process.
+	s.cmd.Env = os.Environ()
+	for k, v := range s.config.Environment {
+		s.cmd.Env = append(s.cmd.Env, k+"="+v)
 	}
-	logBuffer := servicelog.NewRingBuffer(maxLogBytes)
+
+	// Set up stdout and stderr to write to log ring buffer.
 	var outputIterator servicelog.Iterator
-	if m.serviceOutput != nil {
-		outputIterator = logBuffer.TailIterator()
+	if s.manager.serviceOutput != nil {
+		outputIterator = s.logs.HeadIterator(0)
```

Comment:
Heh, not really either. `TailIterator` was fine before, because we weren't keeping logs after the service died or between runs. Now we are keeping logs from previous runs, but we don't want to copy from the beginning of the ring buffer to `serviceOutput` again (that would duplicate the logs), so start iterating from the head when we restart the service.